# MCSD 6227 Master Project 2

Name: Tan Boon Chu

Matrics No.: MCS211006

Project: EDA and Recommender System for Steam Online Gaming Platform

## Data Import and Checking

In [ ]:
# Import PyDrive and associated libraries
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Other libraries
import pandas as pd
import numpy as np

# Authenticate and create the PyDrive client
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [ ]:
# Download a file based on its file ID.
file_id = '1SdwjxbJ6GG4V2P6OiUzUEJRIYDpRsxYE'
downloaded = drive.CreateFile({'id': file_id})

# Save file in Colab memory
downloaded.GetContentFile('steam_clean_data.csv')  

In [ ]:
# Read Data file
df = pd.read_csv('steam_clean_data.csv')
df = pd.DataFrame(df)

# print out number of rows and columns
print('Rows:', df.shape[0])
print('Columns:', df.shape[1])

# View first 5 rows
df.head()

Rows: 28375
Columns: 16


appid                       name         developer  positive  negative  \
0      1             Counter-Strike             Valve    124534      3339   
1      2      Team Fortress Classic             Valve      3318       633   
2      3              Day of Defeat             Valve      3416       398   
3      4         Deathmatch Classic             Valve      1273       267   
4      5  Half-Life: Opposing Force  Gearbox Software      5250       288   

                     owners  average_playtime  median_playtime  price  \
0  10,000,000 .. 20,000,000           17612.0            317.0   9.99   
1   5,000,000 .. 10,000,000             277.0             62.0   4.99   
2   5,000,000 .. 10,000,000             187.0             34.0   4.99   
3   5,000,000 .. 10,000,000             258.0            184.0   4.99   
4   5,000,000 .. 10,000,000             624.0            415.0   4.99   

                                           languages   genre  \
0  English, French, German, Italian, Spanish - Sp...  Action   
1  English, French, German, Italian, Spanish - Sp...  Action   
2  English, French, German, Italian, Spanish - Spain  Action   
3  English, French, German, Italian, Spanish - Sp...  Action   
4                    English, French, German, Korean  Action   

                                detailed_description  Ratings Is_Free English  \
0  Play the world's number 1 online action game. ...   121195    Paid     Yes   
1  One of the most popular online action games of...     2685    Paid     Yes   
2  Enlist in an intense brand of Axis vs. Allied ...     3018    Paid     Yes   
3  Enjoy fast-paced multiplayer gaming with Death...     1006    Paid     Yes   
4  Return to the Black Mesa Research Facility as ...     4962    Paid     Yes   

   download  
0  20000000  
1  10000000  
2  10000000  
3  10000000  
4  10000000

In [ ]:
df.dtypes

appid                     int64
name                     object
developer                object
positive                  int64
negative                  int64
owners                   object
average_playtime        float64
median_playtime         float64
price                   float64
languages                object
genre                    object
detailed_description     object
Ratings                   int64
Is_Free                  object
English                  object
download                  int64
dtype: object

In [ ]:
df.isnull().sum()

appid                   0
name                    0
developer               0
positive                0
negative                0
owners                  0
average_playtime        0
median_playtime         0
price                   0
languages               0
genre                   0
detailed_description    0
Ratings                 0
Is_Free                 0
English                 0
download                0
dtype: int64

In [ ]:
df[df.duplicated(keep=False)]

Empty DataFrame
Columns: [appid, name, developer, positive, negative, owners, average_playtime, median_playtime, price, languages, genre, detailed_description, Ratings, Is_Free, English, download]
Index: []

In [ ]:
df.describe()

appid      positive       negative  average_playtime  \
count   28375.000000  2.837500e+04   28375.000000      28375.000000   
mean    60143.055189  8.748495e+02     186.707419        137.472634   
std     24772.753533  1.831796e+04    4153.111135       1780.315617   
min         1.000000  0.000000e+00       0.000000          0.000000   
25%     40698.500000  6.000000e+00       2.000000          0.000000   
50%     60433.000000  2.200000e+01       8.000000          0.000000   
75%     80137.000000  1.130000e+02      38.000000          0.000000   
max    106946.000000  2.644404e+06  487076.000000     190625.000000   

       median_playtime         price       Ratings      download  
count     28375.000000  28375.000000  2.837500e+04  2.837500e+04  
mean        135.571736      7.652513  6.881421e+02  1.714122e+05  
std        2299.763575     11.255978  1.538953e+04  1.733554e+06  
min           0.000000      0.000000 -7.738000e+03  2.000000e+04  
25%           0.000000      1.990000  1.000000e+00  2.000000e+04  
50%           0.000000      4.990000  1.100000e+01  2.000000e+04  
75%           0.000000      9.990000  6.400000e+01  5.000000e+04  
max      190625.000000    595.990000  2.242091e+06  2.000000e+08

Reduce the size of dataset due to hardware limitation

In [ ]:
df1 = sorted(df.positive, reverse=True)

In [ ]:
# number of download at 5000th row
df1[18000]

11

In [ ]:
df = df.loc[(df['download'] >= 20000) & (df['positive'] >= 11)].copy()

In [ ]:
df.reset_index(inplace=True)

In [ ]:
print('Rows:', df.shape[0])
print('Columns:', df.shape[1])
df.head()

Rows: 18316
Columns: 17


index  appid                       name         developer  positive  \
0      0      1             Counter-Strike             Valve    124534   
1      1      2      Team Fortress Classic             Valve      3318   
2      2      3              Day of Defeat             Valve      3416   
3      3      4         Deathmatch Classic             Valve      1273   
4      4      5  Half-Life: Opposing Force  Gearbox Software      5250   

   negative                    owners  average_playtime  median_playtime  \
0      3339  10,000,000 .. 20,000,000           17612.0            317.0   
1       633   5,000,000 .. 10,000,000             277.0             62.0   
2       398   5,000,000 .. 10,000,000             187.0             34.0   
3       267   5,000,000 .. 10,000,000             258.0            184.0   
4       288   5,000,000 .. 10,000,000             624.0            415.0   

   price                                          languages   genre  \
0   9.99  English, French, German, Italian, Spanish - Sp...  Action   
1   4.99  English, French, German, Italian, Spanish - Sp...  Action   
2   4.99  English, French, German, Italian, Spanish - Spain  Action   
3   4.99  English, French, German, Italian, Spanish - Sp...  Action   
4   4.99                    English, French, German, Korean  Action   

                                detailed_description  Ratings Is_Free English  \
0  Play the world's number 1 online action game. ...   121195    Paid     Yes   
1  One of the most popular online action games of...     2685    Paid     Yes   
2  Enlist in an intense brand of Axis vs. Allied ...     3018    Paid     Yes   
3  Enjoy fast-paced multiplayer gaming with Death...     1006    Paid     Yes   
4  Return to the Black Mesa Research Facility as ...     4962    Paid     Yes   

   download  
0  20000000  
1  10000000  
2  10000000  
3  10000000  
4  10000000

In [ ]:
df.drop(columns=['index', 'appid'], inplace= True)

## TF-IDF Text Vectorization and Transformation

Text cleaning

In [ ]:
df['detailed_description'] = [i.replace('\n', ' ') for i in df['detailed_description']]

In [ ]:
# Remove html tags, new line, and carriage return in description
import re
def remove_tag(html_text):
  re_html = re.compile(r'<[^>]+>|【】|\r|\n|\t|http\S+')
  return re_html.sub(' ',html_text)

In [ ]:
df['detailed_description'] = df['detailed_description'].apply(remove_tag)

In [ ]:
df['detailed_description'] = [i.replace("'", '') for i in df['detailed_description']]

In [ ]:
df['detailed_description'] = [i.replace(",", '') for i in df['detailed_description']]

In [ ]:
import string 
df['detailed_description'] = [i.lower().translate(str.maketrans(string.punctuation, ' '*len(string.punctuation))) for i in df['detailed_description']]

In [ ]:
df['detailed_description'] = [i.replace("—", ' ') for i in df['detailed_description']]

In [ ]:
df.detailed_description

0        play the worlds number 1 online action game  e...
1        one of the most popular online action games of...
2        enlist in an intense brand of axis vs  allied ...
3        enjoy fast paced multiplayer gaming with death...
4        return to the black mesa research facility as ...
                               ...                        
18311     a legionarys life  lets you play as a roman s...
18312     story xu sisi is a young     lady who’s looki...
18313    art is fraught with mysteries  pictures are be...
18314     legend says the blacksmith who wants to creat...
18315    in the game you control the spacecraft and in ...
Name: detailed_description, Length: 18316, dtype: object

In [ ]:
df['detailed_description'].iloc[8519]

' trading cards available       about the game investigator is a first person exploration game with horror elements where the events take place in a world of extreme weather cataclysms  the main character used to be a private investigator but now he is but a mere stranger  forced to roam from house to house in search of food and an overnight stay      exploration game    immersive audio effects guarantee unforgettable game experience    various locations starting from mysterious villages to dark basements    intense atmosphere throughout the game    the use of cryengine makes the picture realistic and smooth     free story dlc in the main menu   '

In [ ]:
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

In [ ]:
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()

In [ ]:
w_tokenizer = nltk.tokenize.WhitespaceTokenizer()

In [ ]:
def lemmatize_text(text):
    return [lemmatizer.lemmatize(word, 'v') for word in w_tokenizer.tokenize(text)]

In [ ]:
df['detailed_description'] = df['detailed_description'].apply(lemmatize_text)

In [ ]:
df['detailed_description']

0        [play, the, worlds, number, 1, online, action,...
1        [one, of, the, most, popular, online, action, ...
2        [enlist, in, an, intense, brand, of, axis, vs,...
3        [enjoy, fast, pace, multiplayer, game, with, d...
4        [return, to, the, black, mesa, research, facil...
                               ...                        
18311    [a, legionarys, life, let, you, play, as, a, r...
18312    [story, xu, sisi, be, a, young, lady, who’s, l...
18313    [art, be, fraught, with, mysteries, picture, b...
18314    [legend, say, the, blacksmith, who, want, to, ...
18315    [in, the, game, you, control, the, spacecraft,...
Name: detailed_description, Length: 18316, dtype: object

In [ ]:
df.detailed_description = df.detailed_description.str.join(' ')

In [ ]:
#Import TfIdfVectorizer from scikit-learn
from sklearn.feature_extraction.text import TfidfVectorizer

#Define a TF-IDF Vectorizer Object. Remove all english stop words such as 'the', 'a'
tfidf = TfidfVectorizer(stop_words='english')

#Construct the required TF-IDF matrix by fitting and transforming the data
tfidf_matrix = tfidf.fit_transform(df['detailed_description'])

#Output the shape of tfidf_matrix
tfidf_matrix.shape

(18316, 77526)

Genre-Based 

In [ ]:
df['feature'] = df['genre'] + ' ' + df ['name'] + ' ' + df['developer']

In [ ]:
df['feature'] = [i.lower().translate(str.maketrans('', '', string.punctuation)) for i in df['feature']]

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

count = CountVectorizer(stop_words='english')
count_matrix = count.fit_transform(df['feature'])

## Cosine Similarities 

calculating the dot product will directly give us the cosine similarity score. Therefore, sklearn's linear_kernel() instead of cosine_similarities() since it is faster.

Description-based Recommender

In [ ]:
# Import linear_kernel
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

In [ ]:
#Construct a reverse map of indices and game titles
indices = pd.Series(df.index, index=df['name'])

In [ ]:
# Cosine Similarities Function
def get_recommendations(name, cosine_sim=cosine_sim):
    # index of the game 
    idx = indices[name]
    
    # similarity scores 
    sim_scores = list(enumerate(cosine_sim[idx]))
    
    # Sort based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar game
    sim_scores = sim_scores[1:11]
    print(sim_scores)
    # Get the recommended game indices
    game_indices = [i[0] for i in sim_scores]
    
    # Get the names of the recommended games
    recommended_games = df.iloc[game_indices]['name'].values

    # Add the recommended games as new columns in the DataFrame
    for i, game in enumerate(recommended_games):
        df.loc[df['name']== name, f'recommendation {i + 1}'] = game

    # Return the top 10 most similar games
    return df.loc[game_indices]

In [ ]:
get_recommendations("Team Fortress 2").head(2)

[(3799, 0.20732793359872517), (553, 0.17298164446250955), (115, 0.1689099337359037), (297, 0.16850351690115528), (1, 0.16736714167131833), (1174, 0.16405878166901366), (8768, 0.16347103048945466), (3695, 0.1622886805628202), (9597, 0.1617446696547721), (13781, 0.16112700271544403)]


name             developer  \
3799                                           Gigantic           Motiga Inc.   
553   Red Orchestra 2: Heroes of Stalingrad with Ris...  Tripwire Interactive   

      positive  negative                  owners  average_playtime  \
3799    5065.0    1904.0  2,000,000 .. 5,000,000             140.0   
553    26922.0    2274.0  2,000,000 .. 5,000,000             771.0   

      median_playtime  price                                 languages  \
3799             50.0   0.00                   English, French, German   
553             269.0  19.99  English, French, German, Russian, Polish   

                                          genre  ... recommendation 1  \
3799             Action, Free to Play, Strategy  ...              NaN   
553   Action, Massively Multiplayer, Simulation  ...              NaN   

      recommendation 2 recommendation 3 recommendation 4  recommendation 5  \
3799               NaN              NaN              NaN               NaN   
553                NaN              NaN              NaN               NaN   

     recommendation 6 recommendation 7 recommendation 8 recommendation 9  \
3799              NaN              NaN              NaN              NaN   
553               NaN              NaN              NaN              NaN   

     recommendation 10  
3799               NaN  
553                NaN  

[2 rows x 26 columns]

In [ ]:
df.head()

name         developer  positive  negative  \
0             Counter-Strike             Valve  124534.0    3339.0   
1      Team Fortress Classic             Valve    3318.0     633.0   
2              Day of Defeat             Valve    3416.0     398.0   
3         Deathmatch Classic             Valve    1273.0     267.0   
4  Half-Life: Opposing Force  Gearbox Software    5250.0     288.0   

                     owners  average_playtime  median_playtime  price  \
0  10,000,000 .. 20,000,000           17612.0            317.0   9.99   
1   5,000,000 .. 10,000,000             277.0             62.0   4.99   
2   5,000,000 .. 10,000,000             187.0             34.0   4.99   
3   5,000,000 .. 10,000,000             258.0            184.0   4.99   
4   5,000,000 .. 10,000,000             624.0            415.0   4.99   

                                           languages   genre  ...  \
0  English, French, German, Italian, Spanish - Sp...  Action  ...   
1  English, French, German, Italian, Spanish - Sp...  Action  ...   
2  English, French, German, Italian, Spanish - Spain  Action  ...   
3  English, French, German, Italian, Spanish - Sp...  Action  ...   
4                    English, French, German, Korean  Action  ...   

         recommendation 1           recommendation 2 recommendation 3  \
0   Team Fortress Classic              Mobile Forces   Operation swat   
1  World of Soccer online             Counter-Strike        Head Shot   
2   Day of Defeat: Source  Silent Storm Gold Edition    Front Defense   
3      DmC: Devil May Cry      ROCKETSROCKETSROCKETS      Rocket Wars   
4              Black Mesa      Half-Life: Blue Shift      Half-Life 2   

                     recommendation 4              recommendation 5  \
0                     Jeklynn Heights                Crisis VRigade   
1                        Hero Academy               Battle Carnival   
2  Full Spectrum Warrior: Ten Hammers         Sniper: Ghost Warrior   
3                  Neptune: Arena FPS  Hellblade: Senua's Sacrifice   
4            Half-Life 2: Episode Two                Hero Academy 2   

           recommendation 6 recommendation 7             recommendation 8  \
0             A.I. Invasion   ESport Manager  Front Office Football Eight   
1       Freefall Tournament       StickyBots                        MIGHT   
2    Blitzkrieg 2 Anthology     Blitzkrieg 3                  WWII Online   
3  Avoid - Sensory Overload        Half-Life                Button Frenzy   
4    Alien Breed 2: Assault         Prospekt            The Perfectionist   

                         recommendation 9  \
0                          Hyper Universe   
1                   ThreadSpace: Hyperbol   
2  Strategy & Tactics: Wargame Collection   
3                            Space Rocket   
4                  Alien Breed 3: Descent   

                                 recommendation 10  
0                                    Confrontation  
1                                        Formicide  
2                                   Death to Spies  
3                        Balloon Chair Death Match  
4  Jacob Jones and the Bigfoot Mystery : Episode 2  

[5 rows x 26 columns]

In [ ]:
get_recommendations('Dota 2')

[(17375, 0.25775620438205177), (17264, 0.24365046291528558), (1834, 0.2335848275934546), (6254, 0.2238555184366565), (15836, 0.22290777467531822), (11932, 0.20947470000758922), (16806, 0.20929268083839214), (13212, 0.2077197038768964), (9189, 0.20718603957780998), (785, 0.19414294164135962)]


name                 developer  positive  \
17375               王者英雄 The Great Hero    China Sky Carving Game      14.0   
17264                   Destiny or Fate                   Blaster      32.0   
1834   Heroes of Havoc: Idle Adventures               Kiz Studios      76.0   
6254                         Flamebreak              Nimbly Games     377.0   
15836                Guards of the Gate        Prince Game Studio      18.0   
11932                      Fort Triumph  CookieByte Entertainment      41.0   
16806                     Tower Keepers                Ninja Kiwi      42.0   
13212             Heroes of Hammerwatch                Crackshell    1401.0   
9189                        Shop Heroes           Cloudcade, Inc.    1978.0   
785      Might & Magic: Clash of Heroes            Capybara Games    1293.0   

       negative                owners  average_playtime  median_playtime  \
17375      11.0           0 .. 20,000               0.0              0.0   
17264       7.0           0 .. 20,000               0.0              0.0   
1834      114.0     50,000 .. 100,000               0.0              0.0   
6254       24.0           0 .. 20,000               0.0              0.0   
15836       6.0           0 .. 20,000               0.0              0.0   
11932       8.0           0 .. 20,000               0.0              0.0   
16806      23.0           0 .. 20,000               0.0              0.0   
13212     356.0    100,000 .. 200,000              45.0             45.0   
9189      868.0  500,000 .. 1,000,000            1543.0           2207.0   
785       157.0    200,000 .. 500,000            1595.0           1595.0   

       price                                          languages  \
17375   7.79   English, Simplified Chinese, Traditional Chinese   
17264  14.99                English, Korean, Simplified Chinese   
1834    0.00  English, French, Italian, German, Spanish - Sp...   
6254    9.99                                            English   
15836   9.99                                            English   
11932  19.99                                            English   
16806   0.00  English, French, Italian, German, Spanish - Sp...   
13212  11.99   English, Simplified Chinese, Traditional Chinese   
9189    0.00  English, French, Italian, German, Spanish - Sp...   
785     9.99  English, French, German, Italian, Spanish - Sp...   

                                                   genre  ...  \
17375                Action, RPG, Strategy, Early Access  ...   
17264                 Indie, RPG, Strategy, Early Access  ...   
1834                                   Free to Play, RPG  ...   
6254                                  Action, Indie, RPG  ...   
15836                               Indie, RPG, Strategy  ...   
11932                      Indie, Strategy, Early Access  ...   
16806                                                RPG  ...   
13212                                 Action, Indie, RPG  ...   
9189   Casual, Free to Play, Massively Multiplayer, R...  ...   
785                                        RPG, Strategy  ...   

      recommendation 1  recommendation 2 recommendation 3 recommendation 4  \
17375              NaN               NaN              NaN              NaN   
17264              NaN               NaN              NaN              NaN   
1834               NaN               NaN              NaN              NaN   
6254               NaN               NaN              NaN              NaN   
15836              NaN               NaN              NaN              NaN   
11932              NaN               NaN              NaN              NaN   
16806              NaN               NaN              NaN              NaN   
13212              NaN               NaN              NaN              NaN   
9189               NaN               NaN              NaN              NaN   
785                NaN               NaN              NaN              NaN   

       recommen

In [ ]:
df['detailed_description'].iloc[5383]

'explore five different worlds fight their monstrous inhabitants and try to survive and find an ancient artifact to save the world from destruction for every game session the game will create new environments full of challenge and surprise populate with monsters and ancient magical artifacts there be eight different player class each one design with a unique play style some of them be combat focus others have access to magic spell in 4 different school of magic while others must rely on their special martial skills to survive as you advance through the game you will find over 40 different type of monsters with distinct skills and abilities you will be able to charm them use magic spell and make them join your quest by use some of the ancient magic find in the dungeon you can unlock more advance form of their power and skills the game keep track of your adventure on its online server share the in game events with your friends in real time maintain a leaderboard of the most successful he

In [ ]:
get_recommendations('Counter-Strike: Global Offensive').head(2)

[(5562, 0.27980011225873574), (16176, 0.20576844144061854), (10, 0.18762491671679013), (7, 0.17816066472778794), (20, 0.17209533671996707), (2366, 0.17179098844549523), (9561, 0.15366056331556222), (13192, 0.15255886521429815), (4983, 0.1524318343353018), (4549, 0.15131691432991773)]


name      developer  positive  negative  \
5562   DARIUSBURST Chronicle Saviours        Pyramid     375.0      65.0   
16176                     Cereal Soup  Wild Arts LLC     218.0     239.0   

                  owners  average_playtime  median_playtime  price  \
5562    20,000 .. 50,000              31.0             31.0  49.99   
16176  50,000 .. 100,000               0.0              0.0   0.00   

                                               languages  \
5562   English, Japanese, French, Italian, German, Sp...   
16176                                            English   

                                                   genre  ...  \
5562                                              Action  ...   
16176  Adventure, Free to Play, Massively Multiplayer...  ...   

      recommendation 1  recommendation 2 recommendation 3 recommendation 4  \
5562               NaN               NaN              NaN              NaN   
16176              NaN               NaN              NaN              NaN   

       recommendation 5 recommendation 6 recommendation 7 recommendation 8  \
5562                NaN              NaN              NaN              NaN   
16176               NaN              NaN              NaN              NaN   

      recommendation 9 recommendation 10  
5562               NaN               NaN  
16176              NaN               NaN  

[2 rows x 26 columns]

In [ ]:
get_recommendations('Unturned').head(2)

[(4591, 0.1893742457462107), (10930, 0.1703625307133499), (14426, 0.16505922066707737), (11882, 0.15775125538643597), (1562, 0.15702152357295507), (882, 0.1560847530123434), (10107, 0.15589786511984807), (3831, 0.15448410367588067), (1839, 0.1521365666211454), (2255, 0.14912609135661442)]


name                 developer  positive  \
4591             Romero's Aftermath  Free Reign Entertainment    5668.0   
10930  Stormworks: Build and Rescue          Sunfire Software    1478.0   

       negative                  owners  average_playtime  median_playtime  \
4591     6073.0  2,000,000 .. 5,000,000             188.0             72.0   
10930     171.0       50,000 .. 100,000               0.0              0.0   

       price         languages  \
4591    0.00  English, Russian   
10930  19.99           English   

                                                   genre  ...  \
4591   Violent, Action, Free to Play, Indie, Massivel...  ...   
10930                           Simulation, Early Access  ...   

      recommendation 1  recommendation 2 recommendation 3 recommendation 4  \
4591               NaN               NaN              NaN              NaN   
10930              NaN               NaN              NaN              NaN   

       recommendation 5 recommendation 6 recommendation 7 recommendation 8  \
4591                NaN              NaN              NaN              NaN   
10930               NaN              NaN              NaN              NaN   

      recommendation 9 recommendation 10  
4591               NaN               NaN  
10930              NaN               NaN  

[2 rows x 26 columns]

In [ ]:
get_recommendations("PLAYERUNKNOWN'S BATTLEGROUNDS").head(2)

[(16846, 0.4410932435814774), (17747, 0.44001090039452506), (17866, 0.3975821204699635), (15237, 0.30165086339521363), (14370, 0.28246797597357876), (16204, 0.2700214034144042), (12816, 0.2541579318825133), (17077, 0.2533920412075136), (17303, 0.23412536868800168), (8428, 0.23389624027784653)]


name         developer  positive  negative  \
16846                      PUGNite      OfficialAndy      72.0      54.0   
17747  Mine Royale - Battle Royale  BobTheShoplifter      13.0       5.0   

                 owners  average_playtime  median_playtime  price languages  \
16846  20,000 .. 50,000               0.0              0.0   0.99   English   
17747       0 .. 20,000               0.0              0.0   2.99   English   

                                  genre  ... recommendation 1  \
16846  Action, Adventure, Casual, Indie  ...              NaN   
17747       Action, Adventure, Strategy  ...              NaN   

       recommendation 2 recommendation 3 recommendation 4  recommendation 5  \
16846               NaN              NaN              NaN               NaN   
17747               NaN              NaN              NaN               NaN   

      recommendation 6 recommendation 7 recommendation 8 recommendation 9  \
16846              NaN              NaN              NaN              NaN   
17747              NaN              NaN              NaN              NaN   

      recommendation 10  
16846               NaN  
17747               NaN  

[2 rows x 26 columns]

In [ ]:
get_recommendations('Warframe').head(2)

[(1540, 0.37882620455525284), (12580, 0.22564356040617464), (1541, 0.201280144114539), (14895, 0.17735213453215706), (17290, 0.1556500579466811), (8698, 0.14723893096621718), (8539, 0.14623447408108253), (10708, 0.13089427923077177), (13846, 0.13022770243366769), (1397, 0.1207079667958566)]


name developer  positive  negative                owners  \
1540   Prime World     Nival    1954.0    1378.0  500,000 .. 1,000,000   
12580  Prime Arena     Nival      60.0      37.0           0 .. 20,000   

       average_playtime  median_playtime  price  \
1540               28.0             33.0    0.0   
12580               0.0              0.0    0.0   

                               languages  \
1540   Russian, English, German, Turkish   
12580                   English, Russian   

                                                  genre  ... recommendation 1  \
1540   Action, Free to Play, Massively Multiplayer, RPG  ...              NaN   
12580  Action, Massively Multiplayer, RPG, Early Access  ...              NaN   

       recommendation 2 recommendation 3 recommendation 4  recommendation 5  \
1540                NaN              NaN              NaN               NaN   
12580               NaN              NaN              NaN               NaN   

      recommendation 6 recommendation 7 recommendation 8 recommendation 9  \
1540               NaN              NaN              NaN              NaN   
12580              NaN              NaN              NaN              NaN   

      recommendation 10  
1540                NaN  
12580               NaN  

[2 rows x 26 columns]

In [ ]:
df['detailed_description'].iloc[8292]

'quot just a cleric quot be my new retro style action rpg become the titular wussy cleric himself and set out to avenge your fall comrades by attempt to do what no cleric before you have ever do cleric de commando but del duio you say i can play a quadrillion game just like yours so why should i my answer be simple can you play a quadrillion retro action rpgs of course this isnt some weird sorta jail but can you play a quadrillion game like mine hells no and heres why just a cleric be more than the sum of its part to your average joe youll see retro pixel graphics well at least he draw his own and didnt use realm of the mad gods graphics hub town level gt missions gt get gear amp reward gt upgrade and buy stuff at the town gt repeat game progression meh but hey at least game like that be pretty fun 4 standard stats that you can boost however you see fit oh boy and ill bet 3 of them suck many many different type of gear to find and upgrade normal upgrade go to 4 and then you can bless t

In [ ]:
df.loc[df['name'] == 'Warframe',['detailed_description']].values[0]

array(['just update band together to save the vallis with abilities power by her shield hildryn be always ready to jump back into the fight take control of her devastate abilities as you work together with the community to solve operation bury debts melee change phase one phase 1 of the much anticipate melee revamp be here with these change you’ll experience more fluid combat and a smoother faster bolder system to harness as you slash your way through the origin system phase 1 include uninterrupted combat with no need for lock into melee mode combo resume aim grind slam and more tenno reinforcements become the deceptively sentient warrior with the illusive limbo limina skin customize your ride with new k drive modular part and give your operator a fresh look with one of three new curly hairstyles prime access get prime access now and make the newest prime yours instantly for prime access bundle options click here prime access be a rotate program that include the latest prime warframes 

Genre-based Recommender

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

cosine_sim2 = cosine_similarity(count_matrix, count_matrix)

time: 6.45 s (started: 2023-01-15 12:17:19 +00:00)


In [ ]:
def feature_rec(name, cosine_sim=cosine_sim2):
    idx = indices[name]
    
    sim_scores = list(enumerate(cosine_sim2[idx]))
    
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    sim_scores = sim_scores[1:11]
    print(sim_scores)
    game_indices = [i[0] for i in sim_scores]
    
    return df.loc[game_indices]

time: 1.66 ms (started: 2023-01-15 12:17:26 +00:00)


In [ ]:
feature_rec('Dota 2')

[(19, 0.6666666666666669), (3799, 0.6666666666666669), (4364, 0.6172133998483676), (7748, 0.6123724356957946), (8871, 0.6123724356957946), (1355, 0.5773502691896258), (1768, 0.5773502691896258), (3123, 0.5773502691896258), (3410, 0.5773502691896258), (4307, 0.5773502691896258)]


name         developer  positive  negative  \
19                   Team Fortress 2             Valve    515879     34036   
3799                        Gigantic       Motiga Inc.      5065      1904   
4364                          CroNix            MAGICS       209       244   
7748                         The Lab             Valve      2926       145   
8871  Divided We Fall: Play For Free  KAVA Game Studio        89        76   
1355                   AirBuccaneers    LudoCraft Ltd.       374       252   
1768                        Toribash      Nabi Studios     11143      1563   
3123                    Call to Arms   Digitalmindsoft      5787      2144   
3410                            Boid             Mokus       536       132   
4307                  Games of Glory    Lightbulb Crew       328       254   

                        owners  average_playtime  median_playtime  price  \
19    20,000,000 .. 50,000,000            8495.0            623.0    0.0   
3799    2,000,000 .. 5,000,000             140.0             50.0    0.0   
4364        200,000 .. 500,000              15.0             16.0    0.0   
7748      500,000 .. 1,000,000              95.0            127.0    0.0   
8871               0 .. 20,000               0.0              0.0    0.0   
1355               0 .. 20,000              67.0            125.0    0.0   
1768    2,000,000 .. 5,000,000             130.0             42.0    0.0   
3123      500,000 .. 1,000,000            2550.0            728.0    0.0   
3410        200,000 .. 500,000               5.0              8.0    0.0   
4307        200,000 .. 500,000              74.0            109.0    0.0   

                                              languages  \
19    English, Danish, Dutch, Finnish, French, Germa...   
3799                            English, French, German   
4364  English, French, Italian, German, Spanish - Sp...   
7748  English, French, German, Czech, Finnish, Hunga...   
8871                                            English   
1355                           English, Russian, Polish   
1768  English, French, Portuguese, Russian, Spanish ...   
3123  English, French, Italian, German, Spanish - Sp...   
3410                                 English, Ukrainian   
4307       English, French, German, Portuguese - Brazil   

                                                  genre  \
19                                 Action, Free to Play   
3799                     Action, Free to Play, Strategy   
4364             Action, Casual, Free to Play, Strategy   
7748                                       Free to Play   
8871              Action, Free to Play, Indie, Strategy   
1355         Action, Free to Play, Indie, RPG, Strategy   
1768              Action, Free to Play, Indie, Strategy   
3123  Action, Free to Play, Indie, Simulation, Strategy   
3410      Action, Casual, Free to Play, Indie, Strategy   
4307                     Action, Free to Play, Strategy   

                                   detailed_description  Ratings Is_Free  \
19    the jungle inferno update play the all new jun...   481843    Free   
3799                          new update about the game     3161    Free   
4364  join the battlefield of a post apocalypse worl...      -35    Free   
7748  set in a pocket universe of aperture science t...     2781    Free   
8871  divide we fall be a close combat multiplayer g...       13    Free   
1355  welcome to the ancient north once the cradle o...      122    Free   
1768  quot just when ragdoll be become an irritant f...     9580    Free   
3123  just update enjoy the latest and massive updat...     3643    Free   
3410  boid be real time strategy at its most primal ...      404    Free   
4307  fight alongside your friends on pc and on ps4 ...       74    Free   

     English  download                                            feature  
19       Yes  50000000          action free to play team fortress 2 valve  
3799     Yes   5000000   action free to play st

In [ ]:
feature_rec('Counter-Strike: Global Offensive').head(2)

[(0, 0.5477225575051663), (19, 0.5163977794943223), (22, 0.5163977794943223), (10, 0.4743416490252569), (7748, 0.4743416490252569), (1221, 0.4472135954999579), (17796, 0.4472135954999579), (9195, 0.43412157106222965), (7, 0.4242640687119285), (4146, 0.4242640687119285)]


name developer  positive  negative                    owners  \
0    Counter-Strike     Valve    124534      3339  10,000,000 .. 20,000,000   
19  Team Fortress 2     Valve    515879     34036  20,000,000 .. 50,000,000   

    average_playtime  median_playtime  price  \
0            17612.0            317.0   9.99   
19            8495.0            623.0   0.00   

                                            languages                 genre  \
0   English, French, German, Italian, Spanish - Sp...                Action   
19  English, Danish, Dutch, Finnish, French, Germa...  Action, Free to Play   

                                 detailed_description  Ratings Is_Free  \
0   play the worlds number 1 online action game en...   121195    Paid   
19  the jungle inferno update play the all new jun...   481843    Free   

   English  download                                    feature  
0      Yes  20000000                 action counterstrike valve  
19     Yes  50000000  action free to play team fortress 2 valve

In [ ]:
feature_rec('Unturned').head(2)

[(8203, 0.7378647873726218), (16039, 0.7378647873726218), (10438, 0.6708203932499369), (18281, 0.6708203932499369), (15246, 0.6454972243679029), (15447, 0.6390096504226939), (17536, 0.6390096504226939), (17295, 0.632455532033676), (17625, 0.632455532033676), (3628, 0.6324555320336759)]


name      developer  positive  negative              owners  \
8203            Maui    Kokua Games       152       177  100,000 .. 200,000   
16039  One Ping Only  Triconn Games       168        17    20,000 .. 50,000   

       average_playtime  median_playtime  price languages  \
8203               16.0             31.0    0.0   English   
16039               0.0              0.0    0.0   English   

                                                genre  \
8203   Action, Adventure, Casual, Free to Play, Indie   
16039  Action, Adventure, Casual, Free to Play, Indie   

                                    detailed_description  Ratings Is_Free  \
8203   the gods have trap your sister prove yourself ...      -25    Free   
16039  one ping only be an student build action adven...      151    Free   

      English  download                                            feature  
8203      Yes    200000  action adventure casual free to play indie mau...  
16039     Yes     50000  action adventure casual free to play indie one...

In [ ]:
feature_rec("PLAYERUNKNOWN'S BATTLEGROUNDS").head(2)

[(15714, 0.5892556509887895), (1322, 0.5345224838248487), (3713, 0.4999999999999999), (13976, 0.4999999999999999), (15567, 0.4999999999999999), (15913, 0.4999999999999999), (16167, 0.4999999999999999), (18058, 0.4999999999999999), (10624, 0.49029033784546006), (3319, 0.472455591261534)]


name            developer  positive  negative  \
15714  RUSSIA BATTLEGROUNDS            Agafonoff       789       299   
1322                   DayZ  Bohemia Interactive    118435     77169   

                       owners  average_playtime  median_playtime  price  \
15714      100,000 .. 200,000               0.0              0.0   0.99   
1322   2,000,000 .. 5,000,000            7235.0           2798.0  44.99   

                                               languages  \
15714  English, French, Italian, German, Spanish - Sp...   
1322   English, French, Italian, German, Spanish - Sp...   

                                                   genre  \
15714  Action, Adventure, Indie, Massively Multiplaye...   
1322            Action, Adventure, Massively Multiplayer   

                                    detailed_description  Ratings Is_Free  \
15714  russia battlegrounds be a multiplayer battle r...      490    Paid   
1322   the post soviet country of chernarus be strike...    41266    Paid   

      English  download                                            feature  
15714     Yes    200000  action adventure indie massively multiplayer r...  
1322      Yes   5000000  action adventure massively multiplayer dayz bo...

In [ ]:
feature_rec('Warframe').head(2)

[(466, 0.5477225575051662), (829, 0.5477225575051662), (4146, 0.5477225575051662), (18295, 0.5477225575051662), (1103, 0.545544725589981), (17500, 0.545544725589981), (3377, 0.5163977794943223), (8871, 0.5103103630798288), (19, 0.5000000000000001), (22, 0.5000000000000001)]


name         developer  positive  negative  \
466      Dark Sector  Digital Extremes       337        96   
829  The Darkness II  Digital Extremes     11992      1234   

                     owners  average_playtime  median_playtime  price  \
466       50,000 .. 100,000               0.0              0.0   9.99   
829  1,000,000 .. 2,000,000             191.0            214.0   5.99   

                                             languages   genre  \
466                                    English, French  Action   
829  English, French, German, Italian, Japanese, Sp...  Action   

                                  detailed_description  Ratings Is_Free  \
466  dark sector thrust players into the role of ha...      241    Paid   
829  inspire by the popular comic book series produ...    10758    Paid   

    English  download                                  feature  
466     Yes    100000      action dark sector digital extremes  
829     Yes   2000000  action the darkness ii digital extremes

In [ ]:
df.to_excel('recommender data.xlsx', index=False)

In [ ]:
from google.colab import files
files.download('recommender data.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Evaluation Metrics

To evaluate the recommender system:

To conduct survey of 30 person to rate (1-5) 5 recommendations with picture and description between 2 recommenders, then compare using t test.

In [ ]:
# Download a file based on its file ID.
file_id = '19Ko3NQvlkwARTS_7uwP4AkS2n2VMGx7s'
downloaded = drive.CreateFile({'id': file_id})

# Save file in Colab memory
downloaded.GetContentFile('Responses.csv')  

In [ ]:
dfe = pd.read_csv('Responses.csv')
dfe = pd.DataFrame(dfe)
dfe = dfe[["Ratings 1","Ratings 2"]]
dfe.head()

Ratings 1  Ratings 2
0        3.4        2.8
1        2.4        3.8
2        3.2        3.8
3        3.6        3.4
4        3.8        2.8

In [ ]:
print('No. of Samples:', dfe.shape[0])

No. of Samples: 30


In [ ]:
import scipy.stats as stats

data_group1 = dfe['Ratings 1']
data_group2 = dfe['Ratings 2']

print(np.var(data_group1), np.var(data_group2))

0.5784888888888889 0.45355555555555555


In [ ]:
stats.ttest_ind(a=data_group1, b=data_group2, equal_var=True)

Ttest_indResult(statistic=0.9541624115336206, pvalue=0.3439613610394162)

The p-value (0.344) is greater than alpha (0.05), thus, the null hypothesis of the test is not rejected. We do not have sufficient evidence to say that the ratings between the recommender systems are different.